File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement {
    private MapDirection orientation;
    private Vector2d position;

    public Animal() {
        orientation = MapDirection.NORTH;
        position = new Vector2d(2, 2);
    }

    public Animal(Vector2d position) {
        orientation = MapDirection.NORTH;
        this.position = position;
    }

    @Override
    public String toString() {
        return switch(orientation) {
            case NORTH -> "^";
            case SOUTH -> "v";
            case EAST -> ">";
            case WEST -> "<";
        };
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }

    public void move(MoveDirection direction) {
        Vector2d newPosition = position;
        switch (direction) {
            case RIGHT -> orientation = orientation.next();
            case LEFT -> orientation = orientation.previous();
            case FORWARD -> newPosition = position.add(orientation.toUnitVector());
            case BACKWARD -> newPosition = position.add(orientation.toUnitVector().opposite());
        }

        if (newPosition != position)
            if (0 <= newPosition.getX() && newPosition.getX() <= 4 && 0 <= newPosition.getY() && newPosition.getY() <= 4)
                position = newPosition;
    }

    public void move(MoveDirection direction, MoveValidator moveValidator) {
        Vector2d newPosition = position;
        switch (direction) {
            case RIGHT -> orientation = orientation.next();
            case LEFT -> orientation = orientation.previous();
            case FORWARD -> newPosition = position.add(orientation.toUnitVector());
            case BACKWARD -> newPosition = position.add(orientation.toUnitVector().opposite());
        }

        if (newPosition != position && moveValidator.canMoveTo(newPosition))
            position = newPosition;
    }

    public Vector2d getPosition() {
        return position;
    }

    public MapDirection getOrientation() {
        return orientation;
    }
}

File: proj-lab2/src/main/kotlin/Animal.kt
class Animal(var position: Vector2d, private var orientation: MapDirection = MapDirection.NORTH) {
    override fun toString(): String = when (orientation) {
        MapDirection.NORTH -> "^"
        MapDirection.SOUTH -> "v"
        MapDirection.EAST -> ">"
        MapDirection.WEST -> "<"
    }

    fun isAt(position: Vector2d?): Boolean = this.position == position

    fun move(direction: MoveDirection) {
        var newPosition = position
        when (direction) {
            MoveDirection.RIGHT -> orientation = orientation.next()
            MoveDirection.LEFT -> orientation = orientation.previous()
            MoveDirection.FORWARD -> newPosition = position + orientation.toUnitVector()
            MoveDirection.BACKWARD -> newPosition = position + orientation.toUnitVector().opposite()
        }

        if (newPosition !== position && (newPosition.x in 0..4 && newPosition.y in 0..4))  position = newPosition
    }
}